trl/trainer/sft_trainer.py

```diff
@@ -0,0 +1,244 @@
+            if dataset_text_field is None:
+                raise ValueError(
+                    "You passed `packing=False` to the SFTTrainer, but you didn't pass a `dataset_text_field` argument."
+                )
+
+            if data_collator is None:
+                data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)
+
+            if "max_seq_length" not in dataset_kwargs:
+                warnings.warn(
+                    "You passed `packing=False` to the SFTTrainer, but you didn't pass a `max_seq_length` argument to the `dataset_kwargs`."
+                    "The default value of 512 will be used. And the text will be padded to that value."
+                )
+
+            max_seq_length = dataset_kwargs.get("max_seq_length", 512)
+
+            if train_dataset is not None:
+                train_dataset = self._prepare_non_packed_dataloader(
+                    tokenizer, train_dataset, dataset_text_field, data_collator, max_seq_length
+                )
+            if eval_dataset is not None:
+                eval_dataset = self._prepare_non_packed_dataloader(
+                    tokenizer, eval_dataset, dataset_text_field, data_collator, max_seq_length
+                )
+
+            is_already_dataset = True
+
+        if not is_already_dataset and dataset_text_field is not None:
+            warnings.warn(
+                "You passed a `dataset_text_field` argument to the SFTTrainer. The trainer will automatically take care of creating a `ConstantLengthDataset` for you."
+            )
+            if tokenizer is None:
+                raise ValueError(
+                    "You need to pass a tokenizer when using the SFT Trainer in `dataset_text_field` mode."
+                )
+
+            if train_dataset is not None:
+                train_dataset = ConstantLengthDataset(tokenizer, train_dataset[dataset_text_field], **dataset_kwargs)
+            if eval_dataset is not None:
+                eval_dataset = ConstantLengthDataset(tokenizer, eval_dataset[dataset_text_field], **dataset_kwargs)
+        elif not is_already_dataset and dataset_text_field is None:
+            raise ValueError(
+                "You need to pass a `dataset_text_field` argument to the SFTTrainer if you want to use the `ConstantLengthDataset`."
+            )
+
+        super().__init__(
+            model=model,
+            args=args,
+            data_collator=data_collator,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            tokenizer=tokenizer,
+            model_init=model_init,
+            compute_metrics=compute_metrics,
+            callbacks=callbacks,
+            optimizers=optimizers,
+            preprocess_logits_for_metrics=preprocess_logits_for_metrics,
+        )
+
+    def _prepare_non_packed_dataloader(self, tokenizer, dataset, dataset_text_field, data_collator, max_seq_len):
+        # tokenize the dataset
+        dataset = dataset.map(
```

Comment:
I think you can just tokenize the dataset so you have the input ids. i would not pad at all (the collator will do this, and if a batch has elements that are all shorter it will be faster).

then you can just pass the tokenized dataset along the data collator (`data_collator = DataCollatorForLanguageModeling(tokenizer, mlm=False)`)

we did something similar here: https://huggingface.co/learn/nlp-course/chapter7/6?fw=pt

then we can maybe just call the function `_tokenize_dataset`

Reply:
Thanks for the great pointer!